noise/README.md

```diff
@@ -202,61 +203,96 @@ libp2p components after the handshake is complete and the payload signature has
-
-If `libp2p_data` is non-empty, the `libp2p_data_signature` field MUST contain a
-signature produced with the libp2p identity key. The data to be signed is the
-UTF-8 string `noise-libp2p-early-data:` followed by the contents of the
-`libp2p_data` field.
+The `data` field contains the "early data" provided to the Noise module when
+initiating the handshake, if any. The structure of this data is opaque to
+noise-libp2p and is defined in the connection establishment specs.
 
 Upon receiving the handshake payload, peers MUST decode the public key from the
-`libp2p_key` field into a usable form. The key MUST be used to validate the
-`noise_static_key_signature` field against the static Noise key received in the
-handshake. If the signature is invalid, the connection MUST be terminated
-immediately.
-
-If the `libp2p_data` field is non-empty, the `libp2p_data_signature` MUST be
-validated against the supplied `libp2p_data`. If the signature is invalid, the
-connection MUST be terminated immediately.
-
-If a noise-libp2p implementation does not expose an API for early data, they
-MUST still validate the signature upon receiving a non-empty `libp2p_data`
-field and abort the connection if it is invalid.
+`identity_key` field into a usable form. The key MUST then be used to validate
+the `identity_sig` field against the static Noise key received in the handshake.
+If the signature is invalid, the connection MUST be terminated immediately.
+
+#### Handshake sealing
+
+In the above scheme, the `data` field is not guarded against tampering.
+Consequently, a MITM agent could alter the data in transit, or even record the
+handshake to replay it at a later time with mutated or transplanted data.
+
+An option to counteract such attacks is to sign the data too. However, for
+efficiency and enhanced security purposes, we choose to adopt the [channel
+binding][npf-channel-binding] technique defined in the Noise Protocol Framework.
+
+The Noise Protocol state machine tracks the handshake hash throughout the entire
+handshake exchange. This is a rolling digest of all messages witnessed by each
+peer, and it's exposed by the state machine via an accessor.
+
+Once the Noise handshake has concluded, and the shared secret has been derived,
+each party performs the following actions:
+
+1. Obtains the `HandshakeHash` from the Noise state machine.
+2. Signs the `HandshakeHash` with its libp2p identity key.
+3. Writes the result on the encrypted channel, prefixed with a Protobuf varint
+   length prefix. We call this payload the "handshake seal".
+4. Awaits to receive the handshake seal from its peer.
+5. Verifies the peer's seal signature against the peer's libp2p public key, and
+   its local value of `HandshakeHash`.
+6. If the signature verification passes, the Noise state machine is destroyed
+   and the encrypted channel is handed over to the libp2p stack. If the
+   verification fails, the party terminates the network connection immediately.
```

Comment:
have you considered noise signatures extension - https://github.com/noiseprotocol/noise_sig_spec/blob/master/output/noise_sig.pdf ? 

it replaces dh used for auth (es, se) with signatures over session hash. for example in XX, after ee is mixed into the symmetric state - responder will sign session hash and append signature with sig-key to an output buffer. initiator performs the same in step 3.

extension specifies support only for ed25519, but i guess you can even do RSA, drop noise static key and reuse libp2p static keypair for signatures. though protocol will slightly deviate from the formed specification :)